File: packages/forumcopilot_sdk/lib/models/entities/fc_post.dart
import 'package:dart_mappable/dart_mappable.dart';
import 'package:forumcopilot_sdk/models/mapping/hooks.dart';
import 'fc_attachment.dart';
import 'fc_like.dart';
import 'fc_post_reaction.dart';
import 'fc_post_vote.dart';
import 'fc_thanks.dart';

part 'fc_post.mapper.dart';

/// FCPost (Forum Consolidated Post) is a unified post model for UI consumption
/// that abstracts away the specific implementation details of different post sources.
@MappableClass()
class FCPost with FCPostMappable {
  /// Unique identifier for the post
  String id;

  /// Title of the post
  String title;

  /// Content of the post
  String content;

  /// Topic ID this post belongs to
  String topicId;

  /// Title of the topic this post belongs to (used in search results)
  String? topicTitle;

  /// Author ID of the post creator
  String authorId;

  /// Author name of the post creator
  String authorName;

  /// Author user type (e.g., admin, moderator, user)
  String? authorUserType;

  /// Date/time when the post was created
  @MappableField(hook: MillisOrIsoDateHook())
  DateTime? timestamp;

  /// URL to the author's avatar/icon
  String? authorIconUrl;

  /// Indicates if the post author is currently online
  bool isAuthorOnline;

  /// Indicates if the user can edit this post
  bool canEdit;

  /// Indicates if smilies are allowed in this post
  bool allowSmilies;

  /// List of attachments in this post
  List<FCAttachment> attachments;

  /// List of inline attachments in this post
  List<FCAttachment> inlineAttachments;

  /// List of thanks information for this post
  List<FCThanks> thanksInfo;

  /// List of likes information for this post
  List<FCLike> likesInfo;

  /// Number of likes/main-reactions on this post. Prefer this over
  /// likesInfo.length — likesInfo may be empty when the actor list
  /// hasn't been fetched.
  int likeCount;

  /// POst number of this post in the forum
  int? postNumber;

  /// Moderation capabilities
  bool canBan;
  bool canDelete;
  bool canApprove;
  bool canMove;
  bool canReport;

  /// Moderation statuses
  bool isBanned;
  bool isDeleted;
  bool isApproved;

  /// Social capabilities
  bool isLiked;
  bool isThanked;
  bool canLike;
  bool canThank;

  /// Whether the current user has bookmarked this post (Discourse:
  /// `bookmarked:true`). XF-flavored impls leave false.
  bool bookmarked;

  /// Whether this post is the accepted answer for its topic (Discourse:
  /// `accepted_answer:true`, requires the discourse-solved plugin).
  bool isSolution;

  /// Whether the current viewer can mark this post as the accepted
  /// answer so the UI can render an "Accept this answer" action.
  bool canAcceptAnswer;

  /// Emoji reactions on this post (Discourse: `discourse-reactions`
  /// plugin). Empty when the plugin isn't installed or no one has
  /// reacted.
  List<FCPostReaction> reactions;

  /// Q&A vote state on this post (Discourse: `discourse-post-voting`
  /// plugin). Null on posts in topics without voting - the UI uses
  /// null to hide the vote arrows entirely.
  FCPostVote? vote;

  /// Revision count of this post (Discourse: `version`). A value
  /// greater than 1 means the post has been edited. Null when the
  /// backend doesn't expose revision counts.
  int? editVersion;

  /// Whether this post is a wiki post that any sufficiently trusted
  /// user may edit (Discourse: `wiki`). Backends without a wiki
  /// concept leave false.
  bool isWiki;

  FCPost(
      {required this.id,
      required this.title,
      required this.content,
      required this.topicId,
      this.topicTitle,
      required this.authorId,
      required this.authorName,
      this.authorUserType,
      this.timestamp,
      this.authorIconUrl,
      this.isAuthorOnline = false,
      this.canEdit = false,
      this.allowSmilies = true,
      this.attachments = const [],
      this.inlineAttachments = const [],
      this.thanksInfo = const [],
      this.likesInfo = const [],
      this.likeCount = 0,
      this.postNumber,
      // Moderation capabilities
      this.canBan = false,
      this.canDelete = false,
      this.canApprove = false,
      this.canMove = false,
      this.canReport = false,
      // Moderation statuses
      this.isBanned = false,
      this.isDeleted = false,
      this.isApproved = true,

      /// Social capabilities
      this.isLiked = false,
      this.isThanked = false,
      this.canLike = false,
      this.canThank = false,
      this.bookmarked = false,
      this.isSolution = false,
      this.canAcceptAnswer = false,
      this.reactions = const [],
      this.vote,
      this.editVersion,
      this.isWiki = false});
}

File: plugins/FC_XenForo2/upload/src/addons/ForumCopilot/Entity/FCPost.php
<?php

namespace ForumCopilot\Entity;

/**
 * ForumCopilot Post Entity
 * Forum-agnostic post representation that can be reused across different forum plugins
 */
class FCPost
{
    public $id;
    public $topicId;
    public $authorId;
    public $authorName;
    public $authorUserType;
    public $timestamp;
    public $content;
    public $title;
    public $authorIconUrl;
    public $canEdit;
    public $canDelete;
    public $canReport;
    public $canLike;
    public $isLiked;
    public $likeCount;
    public $attachments;
    public $isApproved;
    public $isDeleted;
    public $isFirstPost;
    public $position;
    public $likesInfo;
    public $isBanned;

    public function __construct($data = [])
    {
        $this->id = $data['id'] ?? '';
        $this->topicId = $data['topicId'] ?? '';
        $this->authorId = $data['authorId'] ?? '';
        $this->authorName = $data['authorName'] ?? '';
        $this->authorUserType = $data['authorUserType'] ?? '';
        $this->timestamp = $data['timestamp'] ?? null;
        $this->content = $data['content'] ?? ($data['postContent'] ?? ''); // Support both for migration
        $this->title = $data['title'] ?? ($data['postTitle'] ?? ''); // Support both for migration
        $this->authorIconUrl = $data['authorIconUrl'] ?? null;
        $this->canEdit = $data['canEdit'] ?? false;
        $this->canDelete = $data['canDelete'] ?? false;
        $this->canReport = $data['canReport'] ?? false;
        $this->canLike = $data['canLike'] ?? false;
        $this->isLiked = $data['isLiked'] ?? false;
        $this->likeCount = $data['likeCount'] ?? 0;
        $this->attachments = $data['attachments'] ?? [];
        $this->isApproved = $data['isApproved'] ?? false;
        $this->isDeleted = $data['isDeleted'] ?? false;
        $this->isFirstPost = $data['isFirstPost'] ?? false;
        $this->position = $data['position'] ?? 0;
        $this->likesInfo = $data['likesInfo'] ?? [];
        $this->isBanned = $data['isBanned'] ?? false;
    }

    public function toArray()
    {
        // Convert attachments to arrays if they are FCAttachment objects
        $attachmentsArray = [];
        if (is_array($this->attachments)) {
            foreach ($this->attachments as $attachment) {
                if ($attachment instanceof FCAttachment) {
                    $attachmentsArray[] = $attachment->toArray();
                } else {
                    $attachmentsArray[] = $attachment;
                }
            }
        } else {
            $attachmentsArray = $this->attachments;
        }

        return [
            'id' => $this->id,
            'topicId' => $this->topicId,
            'authorId' => $this->authorId,
            'authorName' => $this->authorName,
            'authorUserType' => $this->authorUserType,
            'timestamp' => $this->timestamp,
            'content' => $this->content,
            'title' => $this->title,
            'authorIconUrl' => $this->authorIconUrl,
            'canEdit' => $this->canEdit,
            'canDelete' => $this->canDelete,
            'canReport' => $this->canReport,
            'canLike' => $this->canLike,
            'isLiked' => $this->isLiked,
            'likeCount' => $this->likeCount,
            'attachments' => $attachmentsArray,
            'isApproved' => $this->isApproved,
            'isDeleted' => $this->isDeleted,
            'isFirstPost' => $this->isFirstPost,
            'position' => $this->position,
            'likesInfo' => $this->likesInfo,
            'isBanned' => $this->isBanned,
        ];
    }
}
